src/desktop/space.rs

```diff
@@ -399,6 +399,33 @@ impl Space {
         }
     }
 
+    /// Automatically calls `Window::refresh` for the window that belongs to the given surface,
+    /// if managed by this space.
+    pub fn commit(&mut self, surface: &WlSurface) {
+        if is_sync_subsurface(surface) {
+            return;
+        }
+        let found = RefCell::new(None);
+        with_surface_tree_upward(
+            surface,
+            (),
+            |_, _, _| TraversalAction::SkipChildren,
+            |surface, _, _|
+                for window in self.windows() {
+                    if let Some(w) = window.toplevel().get_surface() {
+                        if w == surface {
+                            *found.borrow_mut() = Some(window);
+                            return;
+                        }
+                    }
+                },
+            |_, _, _| found.borrow().is_some()
+        );
```

Comment:
I don't think this logic does what you want, at all. There is no point in using `with_surface_tree` here.

It should be something like:

```rust
let mut root = surface.clone();
while let Some(parent) = get_parent(&root) {
    root = parent;
}

let if let Some(window) = self.windows().iter().find(|w| w.toplevel().get_surface() == Some(&root)) {
    window.refresh();
}

```

Reply:
yeah, I have overcomplicated this a bit...

Reply:
It's more than overcomplicated, the `with_surface_tree_*` only go through the subtree whose given surface is the root, so thay don't allow finding to actual root of the subsurface tree, which is what you need here.